enh: order classes??

Wondering if you can add the option of ordering/sorting class objects in files - I have got into the (bad) habit of placing several related classes inside a single dart file, but it would now be nice to see them ordered as part of the stylizing ??

MacOS

I tried with --config and got this as the output...

```
$ ../flutter-stylizer --config ~/flutter-stylizer.yaml -w ./lib/...
Using config file: /Users/gslender/flutter-stylizer.yaml
2022/06/23 13:31:22 Stylizing 68 files...
panic: runtime error: slice bounds out of range [:24191] with length 10460

goroutine 1 [running]:
github.com/gmlewis/go-flutter-stylizer/dart.(*Client).rewriteClasses(0xc00010fad0?, {0xc0001c8a80?, 0x6300?}, {0xc000279520, 0x3, 0xc0001679c0?})
	/home/glenn/go/src/github.com/gmlewis/go-flutter-stylizer/dart/rewrite.go:244 +0x110
github.com/gmlewis/go-flutter-stylizer/dart.(*Client).StylizeFile(0xc00010fad0, {0xc00014e2b8, 0x18})
	/home/glenn/go/src/github.com/gmlewis/go-flutter-stylizer/dart/dart.go:125 +0x37f
github.com/gmlewis/go-flutter-stylizer/cli.rootRunE(0x153f440?, {0xc000126640, 0x1, 0x4?})
	/home/glenn/go/src/github.com/gmlewis/go-flutter-stylizer/cli/cli.go:156 +0xb10
github.com/spf13/cobra.(*Command).execute(0x153f440, {0xc000100060, 0x4, 0x4})
	/home/glenn/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0x153f440)
	/home/glenn/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x39c
github.com/spf13/cobra.(*Command).Execute(...)
	/home/glenn/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
github.com/gmlewis/go-flutter-stylizer/cli.Execute()
	/home/glenn/go/src/github.com/gmlewis/go-flutter-stylizer/cli/cli.go:177 +0x25
main.main()
	/home/glenn/go/src/github.com/gmlewis/go-flutter-stylizer/cmd/flutter-stylizer/main.go:22 +0x17
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,11 @@ this file.
 Please make sure you have a backup (preferably in git) of your code before running
 "Flutter Stylizer" in case it doesn't handle your code properly.
 
+## [0.1.12] - 2022-06-22
+
+- Add new option [issue #8](https://github.com/gmlewis/go-flutter-stylizer/issues/8):
+  - `sortClassesWithinFile` (default: `false`)
+
 ## [0.1.11] - 2022-06-22
 
 - Fix [issue #6](https://github.com/gmlewis/go-flutter-stylizer/issues/6).

--- a/README.md
+++ b/README.md
@@ -78,6 +78,7 @@ VSCode User Preferences (`Control/Cmd-,`) like this:
       "private-other-methods",
       "build-method",
     ],
+    "sortClassesWithinFile": false,
     "sortOtherMethods": false,
   }
 ```
@@ -105,6 +106,11 @@ As of `v0.1.8`, a new option flag was added:
 - `groupAndSortVariableTypes` (default: `false`)
   - Whether to group public variables separately by type and sort
     them within their groups. Types are: "final", "optional" (`?`), and "normal".
+
+As of `v0.1.12`, a new option flag was added:
+
+- `sortClassesWithinFile` (default: `false`)
+  - Whether to sort multiple classes within each file.
 
 These features are experimental and should be used with caution.
 Please file any bugs you find on the [GitHub issue tracker].
@@ -184,6 +190,11 @@ incorporate the fix into the plugin.
   supported even though the Dart compiler can handle it.
 
 ## Release Notes
+
+### v0.1.12
+
+- Add new option [issue #8](https://github.com/gmlewis/go-flutter-stylizer/issues/8):
+  - `sortClassesWithinFile` (default: `false`)
 
 ### v0.1.11
 

--- a/package.json
+++ b/package.json
@@ -4,7 +4,7 @@
   "displayName": "flutter-stylizer",
   "description": "Flutter Stylizer organizes your Flutter classes in an opinionated and consistent manner.",
   "icon": "images/flutter-stylizer-logo-128x128.png",
-  "version": "0.1.11",
+  "version": "0.1.12",
   "publisher": "gmlewis-vscode",
   "repository": "https://github.com/gmlewis/flutter-stylizer",
   "engines": {
@@ -34,6 +34,11 @@
             "default": false,
             "description": "Whether to group getters separately (before 'public-other-methods') and sort them within their group. (EXPERIMENTAL - use with caution.)"
           },
+          "groupAndSortVariableTypes": {
+            "type": "boolean",
+            "default": false,
+            "description": " Whether to group public variables separately by type and sort them within their groups. Types are: `final`, `optional` (`?`), and `normal`. (EXPERIMENTAL - use with caution.)"
+          },
           "memberOrdering": {
             "type": "array",
             "default": [
@@ -48,6 +53,11 @@
               "build-method"
             ],
             "description": "Ordered list of members to control reordering of source code."
+          },
+          "sortClassesWithinFile": {
+            "type": "boolean",
+            "default": false,
+            "description": "Whether to sort multiple classes within each file. (EXPERIMENTAL - use with caution.)"
           },
           "sortOtherMethods": {
             "type": "boolean",

--- a/src/dart/dart.ts
+++ b/src/dart/dart.ts
@@ -15,11 +15,13 @@ limitations under the License.
 */
 
 import { Class } from './class'
+import { Editor } from './editor'
 import { Entity } from './entity'
 import { isComment } from './line'
 
 // Edit represents an edit of an editor buffer.
 export interface Edit {
+  dc: Class,
   startPos: number,
   endPos: number,
   text: string,
@@ -29,6 +31,7 @@ export interface Options {
   GroupAndSortGetterMethods?: boolean,
   GroupAndSortVariableTypes?: boolean,
   MemberOrdering?: string[],
+  SortClassesWithinFile?: boolean,
   SortOtherMethods?: boolean,
   SeparatePrivateMethods?: boolean,
   Verbose?: boolean,
@@ -49,20 +52,24 @@ export const defaultMemberOrdering = [
 ]
 
 export class Client {
-  constructor(opts: Options | null) {
+  constructor(editor: Editor, opts: Options | null) {
+    this.editor = editor
     this.opts = opts || {
       GroupAndSortGetterMethods: false,
       GroupAndSortVariableTypes: false,
       MemberOrdering: defaultMemberOrdering,
+      SortClassesWithinFile: false,
       SortOtherMethods: false,
       SeparatePrivateMethods: false,
       Verbose: false,
     }
   }
+  editor: Editor
   opts: Options
 
   generateEdits(classes: Class[]): Edit[] {
     const edits: Edit[] = []
+    const allClasses: Edit[] = []
 
     for (let i = classes.length - 1; i >= 0; i--) {
       const dc = classes[i]
@@ -70,20 +77,53 @@ export class Client {
       const endPos = dc.closeCurlyOffset
 
       const [lines, changesMade] = this.reorderClass(dc)
-      if (!changesMade) {
-        continue
-      }
 
       const edit: Edit = {
+        dc: dc,
         startPos: startPos,
         endPos: endPos,
         text: lines.join('\n'),
       }
 
-      edits.push(edit)
+      allClasses.push(edit)
+      if (changesMade) { edits.push(edit) }
+    }
+
+    if (this.opts.SortClassesWithinFile) {
+      return this.sortClassesWithinFile(edits, allClasses)
     }
 
     return edits
+  }
+
+  sortClassesWithinFile(edits: Edit[], allClasses: Edit[]): Edit[] {
+    const isSorted = allClasses.every((v, i, a) => !i || a[i - 1].dc.className > v.dc.className)
+    if (isSorted) { return edits }
+
+    const gt = (a: Edit, b: Edit) => a.dc.className === b.dc.className ? 0 : a.dc.className < b.dc.className ? 1 : -1
+
+    allClasses.sort(gt)
+    const rev = Array.from(allClasses)
+    rev.reverse()
+
+    const result: Edit[] = []
+    for (let i in allClasses) {
+      const cl = allClasses[i]
+      // log.Printf("i=%v, className=%q, old: startPos=%v, endPos=%v, new: startPos=%v, endPos=%v", i, cl.dc.className, cl.startPos, cl.endPos, rev[i].startPos, rev[i].endPos)
+      const csp = this.editor.findClassAbsoluteStart(cl.dc)
+      // log.Printf("i=%v, csp=%v\n%s%s", i, csp,
+      const rcsp = this.editor.findClassAbsoluteStart(rev[i].dc)
+      // log.Printf("i=%v, rcsp=%v\n%s%s", i, rcsp, c.editor.fullBuf[rcsp:rev[i].startPos], rev[i].text)
+
+      result.push({
+        dc: cl.dc,
+        startPos: rcsp,
+        endPos: rev[i].endPos,
+        text: this.editor.fullBuf.substring(csp, cl.startPos) + cl.text,
+      })
+    }
+
+    return result
   }
 
   reorderClass(dc: Class): [string[], boolean] {

--- a/src/dart/editor.ts
+++ b/src/dart/editor.ts
@@ -127,6 +127,17 @@ export class Editor {
     return [this.lines.length, 0]
   }
 
+  // findClassAbsoluteStart finds the very first absolute offset either at the
+  // beginning of the buffer or after the last blank line before a class is defined.
+  findClassAbsoluteStart(dc: Class) {
+    let result = dc.openCurlyOffset
+    for (; result > 0; result--) {
+      if (result > 1 && this.fullBuf.substring(result - 2, result) === "\n\n") { break }
+      if (result > 3 && this.fullBuf.substring(result - 4, result) === "\r\n\r\n") { break }
+    }
+    return result
+  }
+
   // logf logs the line if verbose is true.
   logf(s: string) {
     if (this.verbose) {

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -49,6 +49,7 @@ export function activate(context: vscode.ExtensionContext) {
 
     const groupAndSortGetterMethods = config.get<boolean>('groupAndSortGetterMethods') || false
     const groupAndSortVariableTypes = config.get<boolean>('groupAndSortVariableTypes') || false
+    const sortClassesWithinFile = config.get<boolean>('sortClassesWithinFile') || false
     const sortOtherMethods = config.get<boolean>('sortOtherMethods') || false
 
     const separatePrivateMethods = memberOrdering.includes('private-other-methods')
@@ -60,12 +61,13 @@ export function activate(context: vscode.ExtensionContext) {
       GroupAndSortGetterMethods: groupAndSortGetterMethods,
       GroupAndSortVariableTypes: groupAndSortVariableTypes,
       MemberOrdering: memberOrdering,
+      SortClassesWithinFile: sortClassesWithinFile,
       SortOtherMethods: sortOtherMethods,
       SeparatePrivateMethods: separatePrivateMethods,
     }
 
     const e = new Editor(source, false)
-    const c = new Client(opts)
+    const c = new Client(e, opts)
     const [got, err] = e.getClasses(groupAndSortGetterMethods, separatePrivateMethods)
     if (err !== null) {
       throw Error(err.message)  // Make the compiler happy.

--- a/src/test/suite/dart/class.test.ts
+++ b/src/test/suite/dart/class.test.ts
@@ -32,6 +32,7 @@ export const runParsePhase = (opts: Options | null, source: string, want: Entity
     GroupAndSortGetterMethods: false,
     GroupAndSortVariableTypes: false,
     MemberOrdering: defaultMemberOrdering,
+    SortClassesWithinFile: false,
     SortOtherMethods: false,
     SeparatePrivateMethods: false,
     Verbose: false,
@@ -40,6 +41,7 @@ export const runParsePhase = (opts: Options | null, source: string, want: Entity
     testOpts.GroupAndSortGetterMethods = opts.GroupAndSortGetterMethods
     testOpts.GroupAndSortVariableTypes = opts.GroupAndSortVariableTypes || false
     testOpts.MemberOrdering = opts.MemberOrdering
+    testOpts.SortClassesWithinFile = opts.SortClassesWithinFile
     testOpts.SortOtherMethods = opts.SortOtherMethods
     testOpts.SeparatePrivateMethods = opts.SeparatePrivateMethods
     verbose = opts.Verbose || false
@@ -47,7 +49,7 @@ export const runParsePhase = (opts: Options | null, source: string, want: Entity
 
   const e = new Editor(source, verbose)
 
-  const c = new Client(testOpts)
+  const c = new Client(e, testOpts)
   const [got, err] = e.getClasses(testOpts.GroupAndSortGetterMethods || false, testOpts.SeparatePrivateMethods || false)
   if (err !== null) {
     throw Error(err.message)  // Make the compiler happy.
@@ -1256,5 +1258,168 @@ class Test {
     wc.lines.forEach((line, i) => {
       assert.strictEqual(line.entityType, want[i], `entityType: line #${wzLineOffset + i + 1}: ${line.line}`)
     })
+  })
+
+  test('Multiple classes with comments', () => {
+    const classA = `class ClassA {
+  ClassA();
+}`
+    const classB = `/// ClassB comment1
+/// ClassB comment2
+class ClassB {
+  ClassB();
+}`
+    const classC = `/* ClassC comment1
+   ClassC comment2
+   ClassC comment 3 */
+class ClassC {
+  ClassC();
+}`
+    const classD = `@JsonSerializable()
+class ClassD {
+  ClassD();
+}`
+    const classE = `/// ClassE comment1
+@reflectiveTest
+class ClassE {
+  ClassE();
+}`
+    const abstractClassF = `/// ClassF comment1
+@reflectiveTest
+abstract class ClassF {
+  ClassF();
+}`
+    const classG = `class _ClassG {
+  _ClassG();
+}`
+    const mainFunc = `main() {
+}`
+    const randomStuff = `typedef DisposeHandler = Future Function();
+
+Version _versionFromString(String input) =>
+    input == null ? null : Version.parse(input);`
+
+    const tests = [
+      {
+        name: "sorted classes",
+        parts: [mainFunc, randomStuff, classA, classB, classC, classD, classE, abstractClassF, classG],
+        want: `main() {
+}
+
+typedef DisposeHandler = Future Function();
+
+Version _versionFromString(String input) =>
+    input == null ? null : Version.parse(input);
+
+class ClassA {
+  ClassA();
+}
+
+/// ClassB comment1
+/// ClassB comment2
+class ClassB {
+  ClassB();
+}
+
+/* ClassC comment1
+   ClassC comment2
+   ClassC comment 3 */
+class ClassC {
+  ClassC();
+}
+
+@JsonSerializable()
+class ClassD {
+  ClassD();
+}
+
+/// ClassE comment1
+@reflectiveTest
+class ClassE {
+  ClassE();
+}
+
+/// ClassF comment1
+@reflectiveTest
+abstract class ClassF {
+  ClassF();
+}
+
+class _ClassG {
+  _ClassG();
+}`,
+      },
+      {
+        name: "reversed classes",
+        parts: [
+          mainFunc,
+          randomStuff,
+          classG,
+          abstractClassF,
+          classE,
+          classD,
+          classC,
+          classB,
+          classA,
+        ],
+        want: `main() {
+}
+
+typedef DisposeHandler = Future Function();
+
+Version _versionFromString(String input) =>
+    input == null ? null : Version.parse(input);
+
+class ClassA {
+  ClassA();
+}
+
+/// ClassB comment1
+/// ClassB comment2
+class ClassB {
+  ClassB();
+}
+
+/* ClassC comment1
+   ClassC comment2
+   ClassC comment 3 */
+class ClassC {
+  ClassC();
+}
+
+@JsonSerializable()
+class ClassD {
+  ClassD();
+}
+
+/// ClassE comment1
+@reflectiveTest
+class ClassE {
+  ClassE();
+}
+
+/// ClassF comment1
+@reflectiveTest
+abstract class ClassF {
+  ClassF();
+}
+
+class _ClassG {
+  _ClassG();
+}`,
+      },
+    ]
+
+    const opts = {
+      MemberOrdering: defaultMemberOrdering,
+      SortClassesWithinFile: true,
+    }
+
+    for (let tt of tests) {
+      const source = tt.parts.join("\n\n")
+      runFullStylizer(opts, source, tt.want, [])
+      // Run again to make sure no extra blank lines are added.
+      runFullStylizer(opts, tt.want, tt.want, [])
+    }
   })
 })

